### DISCUSSION TITLE
Protecting github workflow pipelines from triggers and modifications

### DISCUSSION THREAD

**Discussion by Gby56** [community/community#37986] [Actions]
Hello !

I'm currently reading a lot about the hardening of Github workflows, and I went over all the blog articles and documentations discussing pwn request, hardening etc...
One thing that still stands out to me is: how do I protect workflows from running with changes, if all the people on the repo can create branches and make PRs ?

It doesn't seem straightforward to me, how I could ensure that the workflows that run are ALWAYS code that resides on a protected branch.

There is the `pull_request_target` that could help in theory, making sure that some collaborator modifying a workflow in their branch, isn't triggering their potentially malicious pipeline.

In my head, if we want to ensure a fully secured product codebase, AND its supply chain + build process, then the workflows themselves should never run anything other than the workflow code in the protected branch.
A bit like a production line, multiple people should validate changes to a workflow before it even runs, you don't want a weird PR from a collaborator to dump secrets in a weird way, just because they opened a PR.

There is a way to manually run workflows coming from forks I believe, is that possible to have in a single repo, on a PR that tries to modify a workflow from a protected branch ? does CODEOWNERS prevent the run of the modified workflow ?

I did see the opposite request too, https://github.com/orgs/community/discussions/13836 to allow actions to bypass protection rules too. In some cases I agree it makes sense, but I don't think we should let pipelines run wild !

Thanks a lot to whoever is reading this !

(EDIT: also, Github environments could help here in theory, only let workflows access your secrets if the workflow code is from the protected branch, but that's a bit harder to put in place 🤔 I'd prefer disabling workflows running automatically on PRs that modify a workflow, and take the time to review things)

**Comment by mickeygousset** [comment]
I see customers using the CODEOWNERS file as a partial gate, to ensure that certain eyes get on the PR for the changed workflow.

Also, I see customers using branch protection rules to limit WHO can merge to a branch, as a second way of gating.

**Reply by Gby56** [reply]
Gotcha ! Yes I saw that as a "half" measure, because it won't prevent running the changed workflow automatically right ?
I was checking out if there was an equivalent in the Jenkins world and this article clearly highlights the same threat https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-masters/why-jenkinsfile-changes-are-not-reflected-in-pr-build

I definitely think it would be interesting to have a checkbox on the repo, or at the org-level, to only allow running workflows defined on protected branches. CODEOWNERS simply prevents the persistence because they won't get merged, PRs from forks are not a threat because they don't have access to the original secrets, but PRs internally are technically still dangerous, and only the Environments could maybe mitigate that.... but it still would run modified workflows, it just isolates secrets from that.

**Reply by Gby56** [reply]
To be clear: I re-read the jenkins docs, it seems that they mostly are concerned about PRs from forks, same as Github. I don't see any current pipeline system being cautious about not running untrusted code, from branches in the same repository.
I consider that it should be a feature added into Github workflows, since there is a notion of "protected" branches, it should also be possible to only run workflows from that branch imo

**Reply by raja-anbazhagan** [reply]
> I consider that it should be a feature added into Github workflows

Shouldn't it be a configuration from CIs like Jenkins instead of the repository?

**Comment by jsoref** [comment]
So, you don't want to use `pull_request_target` without being very careful.

But a few things to consider:
1. You should set `permissions` as restrictive as possible. 
2. Even if you need `write` permissions for a workflow, that doesn't mean that you need them for all jobs in a workflow. You can and probably should consider splitting the workflow between two jobs, a `read` job and a `write` job.
3. While `actions/checkout` will default to checking out _something_, possibly someone else's code, it doesn't have to. You can write a workflow to always check out `main` or `master` or _any other reference_. 

To see some examples of the above, you can look at:
https://github.com/check-spelling/spell-check-this/blob/3a35a1cec1e660ab04638cbd5bdd086cf2fcd5c5/.github/workflows/spelling.yml#L102

Note that I'm not an expert, and I've made mistakes in this area, so please don't blindly trust what you read from me (or really anyone else).

**Reply by koppor** [reply]
I think, the scenario is that a contributor modifies the workflow file for themselves. Adding new permissions, adding new actions, etc.

I did not find any documentation whether this is prevented somehow.

**Reply by jsoref** [reply]
GitHub's default model assumes that if you don't trust people you won't give them write access to a repository.

If you're using GHES, then you should look into [Enforcing policy with pre-receive hooks](https://docs.github.com/en/enterprise-server@3.9/admin/policies/enforcing-policy-with-pre-receive-hooks).

If you aren't using GHES and don't absolutely trust everyone, then you should set up your private repositories to allow forking and then limit the people w/ write access to your main repositories and require everyone else to use cross-repository PRs.
